File: scripts/consolidate-interlinear.js
/**
 * Consolidate Interlinear TSV Files into JSON
 *
 * Reads all TSV files from the interlinear directory and merges them into
 * a single JSON file with the following structure:
 *
 * {
 *   "chapters": [
 *     {
 *       "number": 1,
 *       "verses": [
 *         {
 *           "number": 1,
 *           "words": [
 *             {"de": "Am", "en": "IN-THE"},
 *             {"de": "Anfang", "en": "BEGINNING"},
 *             ...
 *           ]
 *         }
 *       ]
 *     }
 *   ]
 * }
 */

const fs = require('fs');
const path = require('path');

/**
 * Parse a single TSV file and return verses
 * @param {string} filepath - Path to TSV file
 * @returns {Array} Array of verse objects
 */
function parseTsvFile(filepath) {
  const content = fs.readFileSync(filepath, 'utf-8');
  const lines = content.split('\n');
  const verses = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    // Split by tab
    const parts = trimmed.split('\t');
    if (parts.length !== 2) continue;

    const [ref, wordPairs] = parts;

    // Parse reference (chapter:verse)
    const refMatch = ref.match(/^(\d+):(\d+)$/);
    if (!refMatch) continue;

    const chapter = parseInt(refMatch[1], 10);
    const verseNum = parseInt(refMatch[2], 10);

    // Parse word pairs
    const words = [];
    for (const pair of wordPairs.split(' ')) {
      if (pair.includes('|')) {
        const [de, en] = pair.split('|', 2);
        words.push({ de, en });
      }
    }

    verses.push({
      chapter,
      verse: verseNum,
      words
    });
  }

  return verses;
}

/**
 * Consolidate all TSV files into a single JSON file
 * @param {string} inputDir - Directory containing TSV files
 * @param {string} outputFile - Path to output JSON file
 */
function consolidateFiles(inputDir, outputFile) {
  // Get all TSV files
  const files = fs.readdirSync(inputDir)
    .filter(f => f.endsWith('.tsv'))
    .sort()
    .map(f => path.join(inputDir, f));

  if (files.length === 0) {
    console.log(`No TSV files found in ${inputDir}`);
    return;
  }

  console.log(`Found ${files.length} TSV files`);

  // Parse all files
  const allVerses = [];
  for (const filepath of files) {
    console.log(`  Processing ${path.basename(filepath)}...`);
    const verses = parseTsvFile(filepath);
    allVerses.push(...verses);
  }

  console.log(`Total verses: ${allVerses.length}`);

  // Organize by chapter
  const chaptersDict = {};
  for (const verse of allVerses) {
    const chapterNum = verse.chapter;
    if (!chaptersDict[chapterNum]) {
      chaptersDict[chapterNum] = {
        number: chapterNum,
        verses: []
      };
    }

    chaptersDict[chapterNum].verses.push({
      number: verse.verse,
      words: verse.words
    });
  }

  // Sort chapters and verses
  const chapters = [];
  const chapterNums = Object.keys(chaptersDict).map(n => parseInt(n, 10)).sort((a, b) => a - b);

  for (const chapterNum of chapterNums) {
    const chapter = chaptersDict[chapterNum];
    chapter.verses.sort((a, b) => a.number - b.number);
    chapters.push(chapter);
  }

  // Build output structure
  const output = { chapters };

  // Write JSON
  fs.writeFileSync(outputFile, JSON.stringify(output, null, 2), 'utf-8');
  console.log(`Wrote ${chapters.length} chapters to ${outputFile}`);
}

// Main execution
if (require.main === module) {
  const scriptDir = __dirname;
  const projectDir = path.dirname(scriptDir);

  const inputDir = path.join(projectDir, 'data', 'interlinear');
  const outputFile = path.join(projectDir, 'data', 'genesis_interlinear.json');

  consolidateFiles(inputDir, outputFile);
}

module.exports = { parseTsvFile, consolidateFiles };

File: scripts/consolidate_interlinear.py
#!/usr/bin/env python3
"""
Consolidate Interlinear TSV Files into JSON

Reads all TSV files from the interlinear directory and merges them into
a single JSON file with the following structure:

{
  "chapters": [
    {
      "number": 1,
      "verses": [
        {
          "number": 1,
          "words": [
            {"de": "Am", "en": "IN-THE"},
            {"de": "Anfang", "en": "BEGINNING"},
            ...
          ]
        }
      ]
    }
  ]
}
"""

import os
import json
import glob
import re


def parse_tsv_file(filepath):
    """Parse a single TSV file and return verses."""
    verses = []

    with open(filepath, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue

            # Split by tab
            parts = line.split('\t')
            if len(parts) != 2:
                continue

            ref, word_pairs = parts

            # Parse reference (chapter:verse)
            ref_match = re.match(r'(\d+):(\d+)', ref)
            if not ref_match:
                continue

            chapter = int(ref_match.group(1))
            verse_num = int(ref_match.group(2))

            # Parse word pairs
            words = []
            for pair in word_pairs.split(' '):
                if '|' in pair:
                    de, en = pair.split('|', 1)
                    words.append({'de': de, 'en': en})

            verses.append({
                'chapter': chapter,
                'verse': verse_num,
                'words': words
            })

    return verses


def consolidate_files(input_dir, output_file):
    """Consolidate all TSV files into a single JSON file."""

    # Get all TSV files
    tsv_files = sorted(glob.glob(os.path.join(input_dir, '*.tsv')))

    if not tsv_files:
        print(f"No TSV files found in {input_dir}")
        return

    print(f"Found {len(tsv_files)} TSV files")

    # Parse all files
    all_verses = []
    for filepath in tsv_files:
        print(f"  Processing {os.path.basename(filepath)}...")
        verses = parse_tsv_file(filepath)
        all_verses.extend(verses)

    print(f"Total verses: {len(all_verses)}")

    # Organize by chapter
    chapters_dict = {}
    for verse in all_verses:
        chapter_num = verse['chapter']
        if chapter_num not in chapters_dict:
            chapters_dict[chapter_num] = {
                'number': chapter_num,
                'verses': []
            }

        chapters_dict[chapter_num]['verses'].append({
            'number': verse['verse'],
            'words': verse['words']
        })

    # Sort chapters and verses
    chapters = []
    for chapter_num in sorted(chapters_dict.keys()):
        chapter = chapters_dict[chapter_num]
        chapter['verses'] = sorted(chapter['verses'], key=lambda v: v['number'])
        chapters.append(chapter)

    # Build output structure
    output = {
        'chapters': chapters
    }

    # Write JSON
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f"Wrote {len(chapters)} chapters to {output_file}")


if __name__ == '__main__':
    # Default paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_dir = os.path.dirname(script_dir)

    input_dir = os.path.join(project_dir, 'data', 'interlinear')
    output_file = os.path.join(project_dir, 'data', 'genesis_interlinear.json')

    consolidate_files(input_dir, output_file)
